### PR TITLE
kernel/syscall: add general description of Phoenix-RTOS system calls

### PR DESCRIPTION
**Pull request by HBuczynski**
As title says, general description of system calls with an example presenting procedure of adding new system call.

### CODE REVIEW

File: kernel/syscalls/README.md [hunk 1]
```diff
@@ -1,18 +1,58 @@
 # System calls
 
+System call commonly abbreviated to syscall makes an entry point to execute a specific user program's request to a service from the kernel. The operating system kernel runs in privileged mode to protect sensitive software or hardware parts from other software components. User application executing in unprivileged mode does not have access to protected data. Performing a hardware interrupt or conducting a trap handled by the kernel, the user application can obtain sensitive data from the kernel, e.g. information about all processes running in the operating system.
```
Comment (agkaminski) [inline kernel/syscalls/README.md:3]:
```suggestion
System call, commonly abbreviated to syscall, is an entry point to execute a specific user program's request to a service from the kernel. The operating system kernel runs in a privileged mode to protect the sensitive software or hardware parts from the other software components. The user application executing in a unprivileged mode does not have access to the protected data. Performing a hardware interrupt or conducting a trap is handled by the kernel, the user application can obtain the sensitive data from the kernel, e.g. an information about all processes running in the system.
```

File: kernel/syscalls/README.md [hunk 2]
```diff
@@ -1,18 +1,58 @@
...
 
+## Declaration and definition
 
+In Phoenix-RTOS declarations and definitions of the system calls are located in a `libphoenix` library. The list of the all system calls is placed in `phoenix-rtos-kernel/include/syscalls.h` grouped by the categories.
```
Comment (agkaminski) [inline kernel/syscalls/README.md:7]:
```suggestion
In Phoenix-RTOS declarations and definitions of system calls are located in the `libphoenix` standard library. A list of all system calls is placed in a `phoenix-rtos-kernel/include/syscalls.h` header file, grouped by categories.
```
Also, system calls are rather defined in the kernel header, libphoenix only provides function prototypes

File: kernel/syscalls/README.md [hunk 3]
```diff
@@ -1,18 +1,58 @@
...
 
+The system calls declaration should be placed in the appropriate header file in `libphoenix` library, referring to the syscall's category.
```
Comment (agkaminski) [inline kernel/syscalls/README.md:9]:
```suggestion
System call declarations should be placed in the an appropriate header file in the `libphoenix` standard library, referring to the syscall's category.
```
prototypes really, not declarations. Also, maybe we shouldn't mix "system call" with "syscall"

File: kernel/syscalls/README.md [hunk 4]
```diff
@@ -1,18 +1,58 @@
...
+The system calls definitions are placed in `arch/$(TARGET_SUFF)/syscalls.S` and are created based on syscall's list in preprocessing phase of the compilation. Each definition triggers an exception (e.g. SuperVisor Call - svc instruction for ARM Cortex-M or Cortex-A) with an appropriate syscall identification number handled by the kernel in privileged mode. Moreover, arguments to syscall are put on stack and redirected to handler.
```
Comment (agkaminski) [inline kernel/syscalls/README.md:10]:
```suggestion
System call definitions are placed in the `arch/$(TARGET_SUFF)/syscalls.S` file and are created based on a syscall's list via macro. Each definition triggers an exception (e.g. SuperVisor Call - svc instruction for ARM Cortex-M or Cortex-A) with an appropriate syscall identification number handled by the kernel in the privileged mode. Arguments are passed according to the target platform ABI (Application Binary Interface).
```

File: kernel/syscalls/README.md [hunk 5]
```diff
@@ -1,18 +1,58 @@
...
 
+Handlers' definitions for system calls are located in `phoenix-rtos-kernel/syscalls.c`. Each handler should contain an appropriate return type consistent with the declaration in `libphoenix` and take stack pointer as a function argument. The syscall's parameters can be obtained from the user stack using the definition `GETFROMSTACK(stack ptr, arg type, var, id)`.
```
Comment (agkaminski) [inline kernel/syscalls/README.md:12]:
```suggestion
Handler definitions for system calls are located in the `phoenix-rtos-kernel/syscalls.c` file. Each handler should contain an appropriate return type consistent with the protype in `libphoenix` (in practice `int`) and take the user stack pointer as an argument. The syscall parameters can be obtained from the user stack using a macro `GETFROMSTACK(stack ptr, arg type, var, id)`.
```

File: kernel/syscalls/README.md [hunk 6]
```diff
@@ -1,18 +1,58 @@
...
 
+Phoenix-RTOS in kernel mode has access to user memory maps so that pointer to data in user space can be passed as an argument to system call and used in kernel space.
```
Comment (agkaminski) [inline kernel/syscalls/README.md:14]:
```suggestion
The Phoenix-RTOS in kernel mode has an access to the calling process memory, so the pointer to the data in the user space can be passed as an argument to system call and used in the kernel space.
```

File: kernel/syscalls/README.md [hunk 7]
```diff
@@ -1,18 +1,58 @@
...
 
+## Adding syscall
 
+An example of adding system call is conducted based on `threadsinfo` syscall. The procedure should be performed as follow:
```
Comment (agkaminski) [inline kernel/syscalls/README.md:18]:
```suggestion
An example of adding a system call is conducted based on the `threadsinfo` syscall. The procedure should be performed as follows:
```

File: kernel/syscalls/README.md [hunk 8]
```diff
@@ -1,18 +1,58 @@
...
 
+1. Add syscall's declaration in `libphoenix`, e.g. `libphoenix/inlude/sys/threads.h`:
```
Comment (agkaminski) [inline kernel/syscalls/README.md:20]:
```suggestion
1. Add syscall declaration in the `libphoenix`, e.g. `libphoenix/inlude/sys/threads.h`:
```

File: kernel/syscalls/README.md [hunk 9]
```diff
@@ -1,18 +1,58 @@
...
 
+```C
+    extern int threadsinfo(int n, threadinfo_t *info);
+```
 
+2. Update syscall list in `phoenix-rtos-kernel/include/syscalls.h`:
```
Comment (agkaminski) [inline kernel/syscalls/README.md:26]:
```suggestion
2. Update the syscall list in `phoenix-rtos-kernel/include/syscalls.h`:
```

File: kernel/syscalls/README.md [hunk 10]
```diff
@@ -1,18 +1,58 @@
...
 
+```C
+    ID(threadsinfo) \
```
Comment (agkaminski) [inline kernel/syscalls/README.md:29]: I would add a note about the vfork syscall or - even better - would advise that new syscalls should be added on the end of the list.

File: kernel/syscalls/README.md [hunk 11]
```diff
@@ -1,18 +1,58 @@
...
+```
 
+3. Create syscall's handler in `phoenix-rtos-kernel/syscalls.c`:
```
Comment (agkaminski) [inline kernel/syscalls/README.md:32]:
```suggestion
3. Create a syscall handler in `phoenix-rtos-kernel/syscalls.c`:
```

File: kernel/syscalls/README.md [hunk 12]
```diff
@@ -1,18 +1,58 @@
...
+```C
+    int syscalls_threadsinfo(void *ustack)
+    {
+        int n, i;
+        pid_t ppid;
+        threadinfo_t *info;
 
+        /* Get input arguments from stack */
+        GETFROMSTACK(ustack, int, n, 0);
+        GETFROMSTACK(ustack, threadinfo_t *, info, 1);
```
Comment (agkaminski) [inline kernel/syscalls/README.md:43]: Maybe some explanation about this macro usage? It's not obvious at all (e.g. that the order matter and what does last argument mean).

File: kernel/syscalls/README.md [hunk 13]
```diff
@@ -1,18 +1,58 @@
...
+        for (i = 0; i < n; ++i) {
+            if ((ppid = posix_getppid(info[i].pid)) > 0)
+                info[i].ppid = ppid;
+        }
+
+        return n;
+    }
+```
+
+4. System call can be invoked in user application adding `#include <sys/threads.h>` header. This syscall is used by `ps` applet in `psh` (Phoenix-Shell).
```
Comment (agkaminski) [inline kernel/syscalls/README.md:58]:
```suggestion
4. The system call can be invoked in the user application by adding including `sys/threads.h` header. This example syscall is used by the `ps` applet in the `psh` (Phoenix-Shell).
```

**Review by agkaminski — COMMENTED** [review]
I'm no english linguist, but these are imho suggestions

**Comment by HBuczynski** [comment]
Thanks for the precise review. I have applied the majority of Your suggestions.

File: kernel/syscalls/README.md [hunk 14]
```diff
@@ -1,18 +1,63 @@
 # System calls
 
+System call commonly abbreviated to syscall is an entry point to execute a specific user program's request to a service from the kernel. The operating system kernel runs in a privileged mode to protect a sensitive software or hardware parts from other the software components. A user application executing in a unprivileged mode does not have access to the protected data. Performing a hardware interrupt or conducting a trap handled by the kernel, the user application can obtain sensitive data from the kernel, e.g. an information about all processes running in the system.
```
Comment (agkaminski) [inline kernel/syscalls/README.md:3]:
```suggestion
System call (commonly abbreviated to syscall) is an entry point to execute a specific user program's request to a service from the kernel. The operating system kernel runs in a privileged mode to protect a sensitive software and hardware parts from the other software components. A user application executing in an unprivileged mode does not have access to the protected data. Performing a hardware interrupt or conducting a trap handled by the kernel, the user application can obtain sensitive data from the kernel, e.g. an information about all processes running in the system.
```

File: kernel/syscalls/README.md [hunk 15]
```diff
@@ -1,18 +1,63 @@
...
 
+## Prototypes and definition
 
+In Phoenix-RTOS prototypes and definitions of the the system calls are located in the `libphoenix` library. A list of the all system calls is placed in a `phoenix-rtos-kernel/include/syscalls.h` header files, grouped by categories.
 
+System call prototypes should be placed in the an appropriate header file in the `libphoenix` standard library, referring to the syscall's category.
+System call definitions are placed in the `arch/$(TARGET_SUFF)/syscalls.S` file and are created based on a syscall's list via macro. Each definition triggers an exception (e.g. SuperVisor Call - svc instruction for ARM Cortex-M or Cortex-A) with an appropriate syscall identification number handled by the kernel in the privileged mode. Arguments are passed according to the target platform ABI (Application Binary Interface).
```
Comment (agkaminski) [inline kernel/syscalls/README.md:10]:
```suggestion
System call definitions are placed in the `arch/$(TARGET_SUFF)/syscalls.S` file and are created based on a syscall list via macro. Each definition triggers an exception (e.g. SuperVisor Call - svc instruction for ARM Cortex-M or Cortex-A) with an appropriate syscall identification number handled by the kernel in the privileged mode. Arguments are passed according to the target platform ABI (Application Binary Interface).
```

File: kernel/syscalls/README.md [hunk 16]
```diff
@@ -1,18 +1,63 @@
...
+        int n, i;
+        pid_t ppid;
+        threadinfo_t *info;
 
+        /* Get input arguments from stack according to order in the syscall prototype.
+           GETFROMSTACK arguments:
+            1st arg - stack pointer
+            2nd arg - type of data put on stack
+            3rd arg - variable name
+            4th arg - id order on stack (start from 0) */
```
Comment (agkaminski) [inline kernel/syscalls/README.md:46]: Maybe an information that the order of GETFROMSTACKS matter (as it is modifying value of ustack pointer)?

**Review by agkaminski — APPROVED** [review]

**Pull request merged**